### PR TITLE
Add compute items for GH constraints

### PR DESCRIPTION
**Pull request by prayush**
## Proposed changes

This PR adds compute tags for GH constraints, and their tests. As constraints depend on the gauge source function H and its space/time deriviatives, we also add two compute items to assemble those from 3+1 variables and their derivatives. (These compute tags will be needed when evolving / observing from the GeneralizedHarmonic system.)

Note: This depends on #1468 , #1469 , #1470 , #1471 , #1473 , #1474 , #1475 , #1476 , #1477 , and #1478  , and changes from them have been patched here as the first commit. Please review the most recent commit only.

**Edit:** All dependencies have been merged into develop, and this PR rebased onto them.

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_Constraints.cpp [hunk 1]
```diff
@@ -931,3 +1211,21 @@ SPECTRE_TEST_CASE("Unit.Evolution.Systems.GeneralizedHarmonic.ConstraintEnergy",
   test_constraint_energy_random<3, Frame::Inertial, double>(
       std::numeric_limits<double>::signaling_NaN());
 }
+
+SPECTRE_TEST_CASE(
+    "Unit.Evolution.Systems.GeneralizedHarmonic.ConstraintComputeTags",
+    "[Unit][Evolution]") {
+  pypp::SetupLocalPythonEnvironment local_python_env{
+      "Evolution/Systems/GeneralizedHarmonic/"};
+  // Test the F constraint against Kerr Schild
```
Comment (prayush) [inline tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_Constraints.cpp:1216]: ping @carmaza - how do I avoid this test from timing out?

Comment (carmaza) [reply tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_Constraints.cpp:1216]:
```
// [[TimeOut, 5]]
SPECTRE_TEST_CASE(
    "Unit.Evolution.Systems.GeneralizedHarmonic.ConstraintComputeTags",
    "[Unit][Evolution]") {
```

where 5 is the time in seconds for a timeout.

**Comment by geoffrey4444** [comment]
Conflicts...can you please rebase @prayush?

**Comment by prayush** [comment]
> Conflicts...can you please rebase @prayush?

Will do so as soon as #1471 is merged, which should be soon enough!

**Comment by nilsdeppe** [comment]
@prayush I just merged the last dependency :)

**Comment by prayush** [comment]
@geoffrey4444 @nilsdeppe - rebased onto develop

File: src/Evolution/Systems/GeneralizedHarmonic/Tags.hpp [hunk 2]
```diff
@@ -166,6 +166,53 @@ struct EvolvedFieldsFromCharacteristicFields : db::SimpleTag {
     return "EvolvedFieldsFromCharacteristicFields";
   }
 };
+
+// @{
+/*!
+ * \ingroup GeneralizedHarmonicGroup
```
Comment (nilsdeppe) [inline src/Evolution/Systems/GeneralizedHarmonic/Tags.hpp]: This group doesn't exist

File: src/Evolution/Systems/GeneralizedHarmonic/TagsDeclarations.hpp [hunk 3]
```diff
@@ -43,6 +43,20 @@ template <size_t Dim, typename Frame>
...
+struct FConstraint;
+template <size_t SpatialDim, typename Frame>
+struct TwoIndexConstraint;
+template <size_t SpatialDim, typename Frame>
+struct ThreeIndexConstraint;
+template <size_t SpatialDim, typename Frame>
+struct FourIndexConstraint;
+template <size_t SpatialDim, typename Frame>
+struct ConstraintEnergy;
+// @}
```
Comment (nilsdeppe) [inline src/Evolution/Systems/GeneralizedHarmonic/TagsDeclarations.hpp]: The `// @}` isn't used here

File: src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp [hunk 4]
```diff
@@ -715,5 +722,304 @@ struct TraceExtrinsicCurvatureCompute
       const tnsr::II<DataVector, SpatialDim, Frame>&) = &trace;
   using base = gr::Tags::TraceExtrinsicCurvature<DataVector>;
 };
+
+// @{
+/*!
+ * \ingroup GeneralizedHarmonicGroup
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp]: This group doesn't exist, apply below several times.

File: src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp [hunk 5]
```diff
@@ -715,5 +722,304 @@ struct TraceExtrinsicCurvatureCompute
...
+  using argument_tags = tmpl::list<::Tags::Coordinates<SpatialDim, Frame>>;
+  static auto function(
+      const tnsr::I<DataVector, SpatialDim, Frame>& coords) noexcept {
+    const DataVector r_squared = get(dot_product(coords, coords));
+    Scalar<DataVector> gamma = make_with_value<type>(coords, 0.);
+    get(gamma) = 3. * exp(-0.5 * r_squared / 64.) + 0.001;
+    return gamma;
+  }
+  using base = ConstraintGamma0;
+  using type = typename base::type;
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp]: remove the `type` type aliases

File: src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp [hunk 6]
```diff
@@ -715,5 +722,304 @@ struct TraceExtrinsicCurvatureCompute
...
 };
+
+// @{
+/*!
+ * \ingroup GeneralizedHarmonicGroup
+ * \brief Compute item to compute constraint-damping parameters for a single-BH
+ * evolution.
+ *
+ * \details Can be retrieved using
+ * `GeneralizedHarmonic::Tags::ConstraintGamma0`,
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp:731]: I'm confused why this references the same tag 3 times. Am I misreading something?

File: src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp [hunk 7]
```diff
@@ -715,5 +722,304 @@ struct TraceExtrinsicCurvatureCompute
...
+ * `GeneralizedHarmonic::Tags::ConstraintGamma0`.
+ */
+template <size_t SpatialDim, typename Frame>
+struct ConstraintGamma0Compute : ConstraintGamma0, db::ComputeTag {
+  using argument_tags = tmpl::list<::Tags::Coordinates<SpatialDim, Frame>>;
+  static auto function(
+      const tnsr::I<DataVector, SpatialDim, Frame>& coords) noexcept {
+    const DataVector r_squared = get(dot_product(coords, coords));
+    Scalar<DataVector> gamma = make_with_value<type>(coords, 0.);
+    get(gamma) = 3. * exp(-0.5 * r_squared / 64.) + 0.001;
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp]: You should multiply by `0.5 / 64 = 0.0078125` instead of divide by `64`

File: src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp [hunk 8]
```diff
@@ -715,5 +722,304 @@ struct TraceExtrinsicCurvatureCompute
...
+    return gamma;
+  }
+  using base = ConstraintGamma2;
+  using type = typename base::type;
+};
+// @}
+
+/*!
+ * \ingroup GeneralizedHarmonicGroup
+ * \brief  Compute item to get the implicit gauge source function from 3 + 1
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp:772]: What is the implicit gauge source function? How does this differ from the gauge source function `H_a`?

File: tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_Constraints.cpp [hunk 9]
```diff
@@ -660,17 +661,289 @@ void test_constraint_energy_analytic(const Solution& solution,
...
+  Mesh<3> mesh{grid_size, Spectral::Basis::Legendre,
+               Spectral::Quadrature::GaussLobatto};
+  using Affine = domain::CoordinateMaps::Affine;
+  using Affine3D =
+      domain::CoordinateMaps::ProductOf3Maps<Affine, Affine, Affine>;
+  const auto coord_map =
+      domain::make_coordinate_map<Frame::Logical, Frame::Inertial>(Affine3D{
+          Affine{-1.0, 1.0, lower_bound[0], upper_bound[0]},
+          Affine{-1.0, 1.0, lower_bound[1], upper_bound[1]},
+          Affine{-1.0, 1.0, lower_bound[2], upper_bound[2]},
```
Comment (nilsdeppe) [inline tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_Constraints.cpp]: why the extra comma at the end here?

File: tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_Constraints.cpp [hunk 10]
```diff
@@ -660,17 +661,289 @@ void test_constraint_energy_analytic(const Solution& solution,
...
+
+  // Compute tested quantities locally
+  const auto gamma0 = GeneralizedHarmonic::Tags::ConstraintGamma0Compute<
+      3, Frame::Inertial>::function(x);
+  const auto gamma1 = GeneralizedHarmonic::Tags::ConstraintGamma1Compute<
+      3, Frame::Inertial>::function(x);
+  const auto gamma2 = GeneralizedHarmonic::Tags::ConstraintGamma2Compute<
+      3, Frame::Inertial>::function(x);
+
+  const auto four_idx_constraint =
```
Comment (nilsdeppe) [inline tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_Constraints.cpp]: `idx` -> `index` (we can spare those two extra letters :) )

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
I'm fine with you squashing

File: src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp (hunk 6 above)
Comment (prayush) [reply src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp:731]: that was a typo, fixed :)

Comment (nilsdeppe) [reply src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp:731]: Okay, thanks! :) Makes sense now :)

File: src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp (hunk 8 above)
Comment (nilsdeppe) [reply src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp:772]: Okay, this is the only thing left that I don't understand. I'm not making a change request (yet at least :) ). Could you please explain what you mean? :)

**Review by nilsdeppe — COMMENTED** [review]
Overall I'm happy with this. Once you respond to my one comment I think this will be ready (modulo possibly the one wording change, if necessary :) )

File: src/Evolution/Systems/GeneralizedHarmonic/Tags.hpp [hunk 11]
```diff
@@ -166,6 +166,52 @@ struct EvolvedFieldsFromCharacteristicFields : db::SimpleTag {
...
+ * `ConstraintEnergyCompute` respectively
+ */
+template <size_t SpatialDim, typename Frame>
+struct GaugeConstraint : db::SimpleTag {
+  using type = tnsr::a<DataVector, SpatialDim, Frame>;
+  static std::string name() noexcept { return "GaugeConstraint"; }
+};
+
+template <size_t SpatialDim, typename Frame>
+struct FConstraint : db::SimpleTag {
```
Comment (nilsdeppe) [inline src/Evolution/Systems/GeneralizedHarmonic/Tags.hpp:185]: Hmm, I'm not getting the documentation showing up on anything but the first tag (`GaugeConstraint`). It's also not obvious to me why that is

File: src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp [hunk 12]
```diff
@@ -715,5 +722,285 @@ struct TraceExtrinsicCurvatureCompute
...
+    const DataVector r_squared = get(dot_product(coords, coords));
+    Scalar<DataVector> gamma = make_with_value<type>(coords, 0.);
+    get(gamma) = 3. * exp(-0.0078125 * r_squared) + 0.001;
+    return gamma;
+  }
+  using base = ConstraintGamma0;
+};
+
+template <size_t SpatialDim, typename Frame>
+struct ConstraintGamma1Compute : ConstraintGamma1, db::ComputeTag {
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp:749]: Here too for some reason only the first class is being documented.

File: src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp (hunk 8 above)
Comment (prayush) [reply src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp:772]: Sorry for not getting back on this earlier -

The gauge source function `H_a` is supposed to act as a _source_ term in the evolution equations for lapse `N` and shift `N^k`, i.e.:
```
(∂t − N^k ∂_k)N ~ function of {H_0, H_i, N, N^k}, 
(∂t − N^k ∂_k)N^k ~ another function of {H_i, N, N^k, ...}
```
(see Eq.9-10 of [this paper](https://arxiv.org/pdf/gr-qc/0512093.pdf)). However, instead of adding a sourcing function to the box, `GaugeHImplicitFrom3p1QuantitiesCompute` inverts the above two eqns to get back `{H_0, H_i}` from derivatives of lapse, shift, etc (see [its doc](https://spectre-code.org/group__generalrelativitygroup#ga117a90f0a602f755c96bcb686e988052)). In other words, this compute tag computes the gauge source function that is _implied_ by the current state of evolved 3+1 variables.

To put it another way, this compute item only "gets" the current gauge source function's value that is implicit from evolved 3+1 variables, and cannot be used to "set" it. I thought this distinction is important to highlight in the name of this compute tag since it will insert the simple tag `GaugeH` into databox, which is the same tag that other compute tags, which actually set user-defined values for `H_a` (such as `DampedHarmonicHCompute`), also insert.

File: src/Evolution/Systems/GeneralizedHarmonic/Tags.hpp (hunk 11 above)
Comment (prayush) [reply src/Evolution/Systems/GeneralizedHarmonic/Tags.hpp]: Documentation shows up fine for all constraint tags for me. See this screenshot:

Screenshot from 2019-05-22 13-50-58

File: src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp (hunk 12 above)
Comment (prayush) [reply src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp]: Doc for these also compiles fine for me, see [screenshot here](https://github.com/sxs-collaboration/spectre/pull/1444#discussion_r286612777)

File: src/Evolution/Systems/GeneralizedHarmonic/Tags.hpp (hunk 11 above)
Comment (prayush) [reply src/Evolution/Systems/GeneralizedHarmonic/Tags.hpp:185]: Hmm, this seems to be common for struct documentations, and plausibly has something to do with global Doxy settings.

For e.g., 3 structs are supposed to share documentation in [`src/ApparentHorizons/ComputeItems.hpp`](https://github.com/sxs-collaboration/spectre/blob/develop/src/ApparentHorizons/ComputeItems.hpp#L25), but only for the first one does the documentation actually show up ([this one](https://spectre-code.org/structah_1_1tags_1_1inversespatialmetriccompute)), while for the other two (see [this](https://spectre-code.org/structah_1_1tags_1_1extrinsiccurvaturecompute) and [this](https://spectre-code.org/structah_1_1tags_1_1spatialchristoffelsecondkindcompute)) it doesn't.

On the other hand, when grouping functions together, doxygen works as expected!! For e.g. we have the overloaded [`characteristic_speeds`](https://spectre-code.org/namespacegrmhd_1_1valenciadivclean#a5c22ceb9a14eae9cca29b9aa108b8ca9) functions of GrMhd system, or functions that compute the [`unnormalized_face_normal`](https://spectre-code.org/group__computationaldomaingroup#ga417a2e1563c807ebf515d4e9fa04d425).

Having said this, its not clear to me how to resolve this except to split up documentation for all members of the doxy group.

File: src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp (hunk 12 above)
Comment (prayush) [reply src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp:749]: see [here](https://github.com/sxs-collaboration/spectre/pull/1444#discussion_r286635875)

**Comment by prayush** [comment]
ping @nilsdeppe

File: src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp (hunk 8 above)
Comment (nilsdeppe) [reply src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp:772]: Okay, that makes sense. Thank you for explaining this to me :)

File: src/Evolution/Systems/GeneralizedHarmonic/Tags.hpp (hunk 11 above)
Comment (prayush) [reply src/Evolution/Systems/GeneralizedHarmonic/Tags.hpp:185]: @nilsdeppe As we had discussed, I've resorted to `@copydoc` everywhere

File: src/Evolution/Systems/GeneralizedHarmonic/Tags.hpp [hunk 13]
```diff
@@ -166,6 +166,52 @@ struct EvolvedFieldsFromCharacteristicFields : db::SimpleTag {
     return "EvolvedFieldsFromCharacteristicFields";
   }
 };
+
+// @{
```
Comment (nilsdeppe) [inline src/Evolution/Systems/GeneralizedHarmonic/Tags.hpp]: We no longer need the `// @{`

File: src/Evolution/Systems/GeneralizedHarmonic/Tags.hpp [hunk 14]
```diff
@@ -166,6 +166,52 @@ struct EvolvedFieldsFromCharacteristicFields : db::SimpleTag {
...
+ * `GaugeConstraintCompute`, `FConstraintCompute`, `TwoIndexConstraintCompute`,
+ * `ThreeIndexConstraintCompute`, `FourIndexConstraintCompute`, and
+ * `ConstraintEnergyCompute` respectively
+ */
+template <size_t SpatialDim, typename Frame>
+struct GaugeConstraint : db::SimpleTag {
+  using type = tnsr::a<DataVector, SpatialDim, Frame>;
+  static std::string name() noexcept { return "GaugeConstraint"; }
+};
+//! \copydoc GaugeConstraint
```
Comment (nilsdeppe) [inline src/Evolution/Systems/GeneralizedHarmonic/Tags.hpp]: Let's not introduce another Doxygen syntax and use `/// \copydoc GaugeConstraint` (triple slash instead of double slash + exclamation point) :)

**Review by nilsdeppe — COMMENTED** [review]
Couple comments on the update, please squash the fixes :)

File: src/Evolution/Systems/GeneralizedHarmonic/Tags.hpp (hunk 13 above)
Comment (prayush) [reply src/Evolution/Systems/GeneralizedHarmonic/Tags.hpp]: removed

File: src/Evolution/Systems/GeneralizedHarmonic/Tags.hpp (hunk 14 above)
Comment (prayush) [reply src/Evolution/Systems/GeneralizedHarmonic/Tags.hpp]: changed

**Review by nilsdeppe — APPROVED** [review]
LGTM. Nice job @prayush! :)

**Review by geoffrey4444 — APPROVED** [review]
